update logic surrounding timestepping of cable%cansto / %oldcanst

JhanSrbinovsky opened the pull request:
# CABLE

Thank you for submitting a pull request to the CABLE Project.

## Description

Clarifying time stepping of cansto / oldcansto

Fixes #505

## Type of change

Please delete options that are not relevant.

- [x] Bug fix
- [ ] New or updated documentation



<!-- readthedocs-preview cable start -->
----
📚 Documentation preview 📚: https://cable--506.org.readthedocs.build/en/506/

<!-- readthedocs-preview cable end -->

har917 commented:
@JhanSrbinovsky @ccarouge 

Headline comment:  This looks okay - the science looks correct and it would certainly improve the consistency in logic between the different applications.  It's a sensible half-way stop as part of the broader effort to address #162 #496.

However - one big concerns to check the approach on and one request for change.

concern:  I'm not sure that I like edits to files in the `coupled/CM2` and `coupled/ESM1.5` directories - I would have thought these are now read only.  However, you can't make the edit to `cable_serial` and `cable_canopy` and keep the codebase functional without those changes.  So I'm at a bit of an impass on the way forward.

request for change: we need some comments in the code to describe what's going on better 

----------------------
In CM2/cable_explicit_driver, AM3/cable_explicit_driver and ESM1.5/cable_explicit_driver ahead of the new `%cansto` line add

!cbm/define_canopy() sets the value of %oldcansto then updates the value of %cansto
!we reset to value of %cansto to that at start of timestep here to work with the split structure of the ACCESS calls to CABLE


In cable_canopy ahead of new %cansto line include

!take a copy of the value of %cansto at the start of time step.  This is used to reset %cansto part way through ACCESS timestep
!and in evaluation of some water fluxes.

JhanSrbinovsky commented:
@har917 - hrmmm - you're right, but I have no idea of the way forward. Perhaps, and it will have to happen at some stage, MIPX models of tester-year will just need to be tagged and left alone.

ccarouge commented:
> concern: I'm not sure that I like edits to files in the coupled/CM2 and coupled/ESM1.5 directories 
This is because #474 was merged into src/coupled/ESM1.5 instead of src/coupled/esm16. I've opened issue #507 and asked Ben to sort this out. Changes for ESM1.6 should go into src/coupled/esm16 (which hopefully can be renamed for consistency).

And I have no idea why CM2 need to change? CM2 does not run with CABLE3 so it should not reflect changes for the latest CABLE.

har917 commented:
@JhanSrbinovsky @ccarouge Another option would be to add conditions around `frozen_limit`, %wbice`, `%wb` and `%ssat` that reflect the updated use of the different densities in the initialisation section(s) - i.e. constrain the starting conditions to abide by the updated physics.  The key difference is that we wouldn't have to maintain conservation at that point.

This would, hopefully, avoid the initial temperature shock that you're seeing (which I still think is coming the restart not being in balance with the physics and showing up in `surfbv`).  Given that we're breaking bitwise comparability anyway I don't think it's too much of an issue.

The question is where would it need to go in, presumably
- offline: in between the restart read and first time step
- coupled: in the `init_soil` part of the initialisation routines

Head's up: the work that @rkutteh has been doing (both the work to ensure consistency between soil type and veg type, and the GW work) could well intersect with this awkwardly.

JhanSrbinovsky commented:
@ccarouge - true (I think). I have run CM2 semi-recently but I am not sure of the details. An argument though for just removing the coupled/CM2 directory completely?

@har917  - I'll have to double check but I don't remember having a restart for this GSWP2 run, although GSWP2 was the new kid on the block when I set this up so I might've just forgotten.

ccarouge commented:
@JhanSrbinovsky We need some test results with this. What's the impact on simulations?

JhanSrbinovsky commented:
Ive run this  through  benchcab. Surprisingly reproducibility fails. I can believe a rounding error perhaps, so I will have to plot the data, but my understanding is that it was just nodified so we could use the same time-stepping logic across apps, although the original time-stepping was not wrong per se.

har917 commented:
I'm a bit surprised this has failed on reproducibility given that I thought that we default to using `cansat = 0.0`.   In that case this whole section of code should be effectively redundant (lots of 0.0 = 0.0 lines but no net impact).

However my bigger concern here is that there are changes to the ESM1.5 and CM2 `*driver*` files in this change set - I thought/hoped that we'd agreed to not modify the source code of past applications in the git repos and focus only on AM3 and ESM1.6.

More generally this area of science needs to be reviewed as part of [cansto time stepping](https://github.com/CABLE-LSM/ACCESS3-JULES/issues/74) and the refactoring of the code base.

JhanSrbinovsky commented:
hmmmm - I thought this was the conclusion if the review. Anyway, the coupled changes herein are likely in those apps already. They possibly appeared here before the decision had been made. I'll make sure that the coupled changes are indeed already in those apps and removed them. That shows up how crazy this is that it is not reproducible.  We've moved the single line from outside canopy to inside canopy

har917 commented:
> I'm a bit surprised this has failed on reproducibility given that I thought that we default to using `cansat = 0.0`. In that case this whole section of code should be effectively redundant (lots of 0.0 = 0.0 lines but no net impact).

Correction - it appears that (in offline and ESM1.5 at least) we are using `cansat = veg%canst1 * veg*vlaiw` with `veg%canst1 = 0.1` (irrespective of the surface type!!)

JhanSrbinovsky commented:
@har917 dint we already do this in the Xmas lead up?

har917 commented:
@JhanSrbinovsky Yes we did a lot of work in this.  There are two elements to this topic.

1. changes in ESM1.6/AM3 to sort out the time stepping
2. Hoped for changes to CABLE3 MAIN so that `%cansto` stops being the odd-state-variable-out - see #162 - which would ripple down to ESM1.6 and AM3.

Most of this PR is about 1. and should be in ESM1.6 - but as it's in the coupling code I don't know whether we sync'd back to MAIN - see also #505.  I'd be nervous about merging - especially the difference `cable_canopy`.

JhanSrbinovsky commented:
So is it correct in any of the applications (main/fastrack/AM3)?

har917 commented:
@JhanSrbinovsky @ccarouge I think they are all correct at the moment - but highly confusing and could be rationalized.  What that opinion implies is that this PR isn't needed - except as the means to sync the different repos - but probably worth a second opinion on this.

------------------------------------------------------------
I have that the sequence of operations in offline - as per the CABLE3 MAIN repo - is: 
- %cansto in from restart (if it exists or set in cable_parameters and/or cable_initialize)
- %oldcansto = %cansto in serial driver, mpimaster and mpiworker.
- cbm does nothing
- %cansto = %oldcansto in define_canopy - which then increments %cansto
- %cansto (and %delwc) into balance_checks, restart and output

ESM1.6 (part 1) - as per the UM7 repo:
- %cansto pack'd from canopy_tile in initialize_canopy
- %oldcansto =%cansto in explicit_river
- %oldcansto=%cansto in cbm_expl
- %cansto=%oldcansto in define_canopy - which then increments %cansto
- %cansto = %oldcansto in cbm_expl

then ESM1.6 (part 2)
- %cansto=%oldcansto in implicit_driver
- %cansto=%oldcansto in define_canopy - which then increments %cansto
- canopy_tile unpack'd from %cansto in implicit_unpack

AM3 (part 1) - as per the ACCESS3-JULES repo:
- %cansto pack'd from canopy_tile in um_update
- %oldcansto = %cansto in um_update
- %cansto=%oldcansto in define_canopy - which then increments %cansto
- %cansto = %oldcansto

AM3 (part 2 and 3):
- pb%ocan = %oldcansto in cable_store_prognostics from implicit_main
- %cansto=%oldcansto in define_canopy - which then increments %cansto
- %cansto = %oldcansto *if ***NOT*** the last call each time step*
- %oldcansto = pb%ocan in cable_restore_prognostics from implicit_main
- canopy_tile unpack'd from %cansto in implicit_unpack

---------------------------------------------
If timing weren't an issue - the two immediate changes I'd make/test are
1. move the %oldcansto = %cansto line from serial driver etc. to offline cbm (this brings the cbm's closer together)
2. delete the %cansto=%oldcansto line in define_canopy

In the longer term:
1. sort out and refactor the incrementing of %cansto in define_canopy
2. make other adjustments across the applications 

This would eventually mean that %oldcansto can be removed from the canopy TYPE entirely and %cansto would be handled in the same way as all the other state variables (e.g. in/out of the PB%) and stop being a special case.

JhanSrbinovsky commented:
Thanks @har917. I'm pretty sure we actually put in the casnsto=oldcanstyo in define_canopy across apps as it wasnt the case before. Then the app specific treatment differs
